### DISCUSSION TITLE
Adjustment of auto-import behavior for project-internal imports

### DISCUSSION THREAD

**Discussion by smackesey** [microsoft/pylance-release#3050] [Enhancement]
A symbol can be defined in one module and re-exported from several others. When importing a symbol from an external typed library, pylance's behavior is, according to pyright dev @erictraut:

>If the same name is exported from multiple modules, the auto-import logic in pyright and pylance prefer the shortest module path. Only the shortest path is listed in the completion list, and the longer paths are de-duped.

However, this is not the behavior for internal project-internal imports. AFAICT, for project-internal imports, resolution is to the module that defines the symbol, i.e. usually the _longest_ exporting module path is preferred. This can be undesirable on large projects where you have big submodules that you want to provide an "internal public" interface for-- e.g. if you have a large submodule `project.foo`, you might want project-internal code to import only from `project.foo` itself, rather than `project.foo.*` submodules.

I'd like to suggest that project-internal auto-imports instead resolve to __the shortest non-ancestor module that exposes a symbol__. So:

```
### project/foo/__init__.py
# redundant alias is necessary to expose `BAR` as public for `project.foo`
from .bar import BAR as BAR

### project/foo/bar.py
BAR = "BAR"

### project/baz.py
# "|" is the cursor
BA|  # should offer import "from project.foo import BAR" instead of "from project.foo.bar import BAZ"
```

**Comment by debonte** [comment]
@heejaechang

**Comment by r3m0t** [comment]
Did you try defining `__all__` in `__init__.py`, as that's the usual way to specify what is 'exported' from a module

**Reply by smackesey** [reply]
@r3m0t `__all__` actually does not  does _not_ define a module's public interface (I used to think it did too), it can just be used to make symbols that otherwise would be private public. The full rules are here: https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface
